rfcs/2021-12-01-9460-healthcheck-endpoint-improvements

```diff
@@ -0,0 +1,114 @@
+# RFC 9460 - 2021-12-01 - Healthcheck endpoint improvements
+
+Our existing health endpoint is too limited and additional features are required
+to properly interact with load balancers or Kubernetes service discovery.
+
+## Context
+
+- [Additional response for api health endpoint](https://github.com/vectordotdev/vector/issues/9160)
+
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+
+### Out of scope
+
+- Moving the health endpoint out of our existing API
+- Adding user facing configuration to adjust endpoint behavior or responses
+
+## Pain
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on our GraphQL based API is enhanced to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
```

Comment:
Which behavior in particular, 503 specifically or _not_ responding 200 at all times?

Reply:
Yes, 503s specifically, or how K8s responds to different codes.

Reply:
```
Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure.
```

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request

`503` seemed most appropriate from established status codes.

Reply:
Just noting also that the `/health` route is actually outside of the GraphQL API which is mounted at `/graphql`. This is an important note because I think it means we can use traditional REST semantics here and use status codes to indicate health rather than being restricted by GraphQL which would prefer all requests return 200 and include the health information in the body of the response.

Reply:
Ah, I think I misunderstood some of the organization of the code here. 👍

Reply:
To give a little background on why `/health` is currently coupled with the API:

The code in src/api spins up a warp HTTP server. I added /health originally as a quick route to check for 'aliveness', and followed that up with a `{ health }` equivalent in GraphQL. Both were kept so that a user could choose which is more relevant (the former potentially being simpler to parse, and not requiring any GraphQL knowledge.)

As `/health` evolves to distinguish between aliveness and readiness, I think it ultimately makes sense to do the same in GraphQL. Instead of returning a boolean, we could return an enum with `Ready`, `NotReady`, etc.

That's probably out of scope for this first go, but just something to keep in mind. It'd be useful to reuse the same mechanism for determining health in the GraphQL API.

Reply:
🤔 I'm personally still not sold on the need for both the HTTP and GraphQL based health checks, it feels duplicative to me - but I'm also less familiar with GraphQL and have been reaching for things more like `vector validate` than the GraphQL endpoint as I've envisioned this moving forward. (That also seems easier from the kubernetes frame of reference compared to making GraphQL queries)

Reply:
The GraphQL API provides general observability into a running Vector instance. An argument for including health data is that it complements the other categories of data it already provides: host and internal metrics, and component meta. It also serves a different class of client.

A GraphQL client that is querying for those things could extend a query to include health/status, too. An example use case could be `vector top`; I could see displaying a "Shutting down..." notice in the terminal. By including it in the API, this could even be subscribed to rather than polled.

`/health`, OTOH, feels more suited to general purpose readiness checks.

TL;DR - Two different clients, with typically different needs.